### ISSUE TITLE
does not free wsi totally when getaddrinfo failed in lws_client_connect_2

### ISSUE THREAD

**Issue by freemine** [warmcat/libwebsockets#1187]
as ref to previous discussion on closed-issue: https://github.com/warmcat/libwebsockets/issues/1186

I thought I found the problem occurs in lws.

code snippet as follows:
```
 409 oom4:
 410   /* we're closing, losing some rx is OK */
 411   lws_header_table_force_to_detachable_state(wsi);
 412 
 413   if (wsi->mode == LWSCM_HTTP_CLIENT ||
 414       wsi->mode == LWSCM_HTTP_CLIENT_ACCEPTED ||
 415       wsi->mode == LWSCM_WSCL_WAITING_CONNECT) {
 416     wsi->vhost->protocols[0].callback(wsi,
 417       LWS_CALLBACK_CLIENT_CONNECTION_ERROR,
 418       wsi->user_space, (void *)cce, strlen(cce));
 419     wsi->already_did_cce = 1;
 420   }
 421   /* take care that we might be inserted in fds already */
 422   if (wsi->position_in_fds_table != -1)
 423     goto failed1;
 424   lws_remove_from_timeout_list(wsi);
 425   lws_header_table_detach(wsi, 0);
 426   lws_free(wsi);
 427 
 428   return NULL;
 429 
 430 failed:
 431   wsi->vhost->protocols[0].callback(wsi,
 432     LWS_CALLBACK_CLIENT_CONNECTION_ERROR,
 433     wsi->user_space, (void *)cce, strlen(cce));
 434   wsi->already_did_cce = 1;
 435 failed1:
 436   lws_close_free_wsi(wsi, LWS_CLOSE_STATUS_NOSTATUS, "client_connect2");
 437 
 438   return NULL;
```
difference exists in line 426 and line 436.

Please kindly confirm is this is the case?

Thank you.

**Comment by freemine** [comment]
I tried changing line 426 to:
lws_close_free_wsi(wsi, LWS_CLOSE_STATUS_NOSTATUS, "client_connect2");

And the core-dump due to 'double-free' problem, does not exists.

And LWS_CALLBACK_WSI_DESTROY get called after cce.

**Comment by lws-team** [comment]
With the server side, there is no difference between an incoming connection accept and a valid connected socket, as soon as it was accepted it's valid and working.

But with client connections, the actual socket connection happens asynchronously.  That means for some time after starting the connection, the wsi is not really representing a connected socket.

The idea in that code is that prior to the connect completing, if we decide to fail out, we do it via "oom4", which is a kind of "lite" close.  Past here

```
                /*
                 * past here, we can't simply free the structs as error
                 * handling as oom4 does.  We have to run the whole close flow.
                 */

                if (!wsi->protocol)
                        wsi->protocol = &wsi->vhost->protocols[0];

                wsi->protocol->callback(wsi, LWS_CALLBACK_WSI_CREATE,
                                        wsi->user_space, NULL, 0);

                lws_set_timeout(wsi, PENDING_TIMEOUT_AWAITING_CONNECT_RESPONSE,
                                AWAITING_TIMEOUT);
```

we have the connection and we officially insert the socket fd into the event loop and create the wsi.  After that point in the code, if we error out or close later, we use the regular close api to clean up.  The regular close path notifies the user callback with LWS_CALLBACK_WSI_DESTROY.  So for that and other reasons, we do not call it on the oom4 path.

So while I am happy to see your patch and willing to believe there's a leak, the way the patch is doing it will create a WSI_DESTROY imbalance.

Can you paste what exactly is leaking, from what valgrind told?  The correct way is likely to handle freeing that specific thing at oom4.

**Comment by freemine** [comment]
Thx!

Just a quick response here first.
I checked the code in lws_client_connect_2 again, and I know what you mean. But the situation that
I raised is a little different here.
With the 'http://somewhere.that.does.not.exist', it'll fail at:
```
 106        lwsl_notice("%s: %p: address %s\n", __func__, wsi, ads);
 107 
 108        n = lws_getaddrinfo46(wsi, ads, &result);
 109 
```
and then 
```
 188       lwsl_err("getaddrinfo failed\n");
 189       cce = "getaddrinfo failed";
 190       goto oom4;
```
And thus, neither 
```
 227       wsi->desc.sockfd = socket(AF_INET, SOCK_STREAM, 0);
```
nor
```
 254     lws_change_pollfd(wsi, 0, LWS_POLLIN);
```
was called before jumping to oom4

Difference is here, but I am not sure if this is the case shall be taken care of.



BTW, for the valgrind part, I'll come up with you shortly after.

**Comment by freemine** [comment]
FYI:
```
==30484== 
==30484== HEAP SUMMARY:
==30484==     in use at exit: 216 bytes in 8 blocks
==30484==   total heap usage: 3,394 allocs, 3,386 frees, 438,780 bytes allocated
==30484== 
==30484== 2 bytes in 1 blocks are indirectly lost in loss record 1 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x4C2CF1F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x41813E: _realloc (alloc.c:64)
==30484==    by 0x418186: lws_realloc (alloc.c:77)
==30484==    by 0x41BD43: lws_strdup (client-handshake.c:688)
==30484==    by 0x41C24D: lws_client_connect_via_info (client-handshake.c:808)
==30484==    by 0x40731D: create_request (test-limit.c:376)
==30484==    by 0x4075C4: main (test-limit.c:444)
==30484== 
==30484== 4 bytes in 1 blocks are indirectly lost in loss record 2 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x4C2CF1F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x41813E: _realloc (alloc.c:64)
==30484==    by 0x418186: lws_realloc (alloc.c:77)
==30484==    by 0x41BD43: lws_strdup (client-handshake.c:688)
==30484==    by 0x41C369: lws_client_connect_via_info (client-handshake.c:825)
==30484==    by 0x40731D: create_request (test-limit.c:376)
==30484==    by 0x4075C4: main (test-limit.c:444)
==30484== 
==30484== 30 bytes in 1 blocks are indirectly lost in loss record 3 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x4C2CF1F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x41813E: _realloc (alloc.c:64)
==30484==    by 0x418186: lws_realloc (alloc.c:77)
==30484==    by 0x41BD43: lws_strdup (client-handshake.c:688)
==30484==    by 0x41C22F: lws_client_connect_via_info (client-handshake.c:807)
==30484==    by 0x40731D: create_request (test-limit.c:376)
==30484==    by 0x4075C4: main (test-limit.c:444)
==30484== 
==30484== 30 bytes in 1 blocks are indirectly lost in loss record 4 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x4C2CF1F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x41813E: _realloc (alloc.c:64)
==30484==    by 0x418186: lws_realloc (alloc.c:77)
==30484==    by 0x41BD43: lws_strdup (client-handshake.c:688)
==30484==    by 0x41C26C: lws_client_connect_via_info (client-handshake.c:809)
==30484==    by 0x40731D: create_request (test-limit.c:376)
==30484==    by 0x4075C4: main (test-limit.c:444)
==30484== 
==30484== 30 bytes in 1 blocks are indirectly lost in loss record 5 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x4C2CF1F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x41813E: _realloc (alloc.c:64)
==30484==    by 0x418186: lws_realloc (alloc.c:77)
==30484==    by 0x41BD43: lws_strdup (client-handshake.c:688)
==30484==    by 0x41C2DF: lws_client_connect_via_info (client-handshake.c:815)
==30484==    by 0x40731D: create_request (test-limit.c:376)
==30484==    by 0x4075C4: main (test-limit.c:444)
==30484== 
==30484== 32 bytes in 1 blocks are still reachable in loss record 6 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x50F5DB2: CRYPTO_malloc (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30484==    by 0x517BEAE: sk_new (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30484==    by 0x4E745A9: ??? (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30484==    by 0x4E76528: SSL_COMP_get_compression_methods (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30484==    by 0x4E7B8A8: SSL_library_init (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30484==    by 0x41EAA2: lws_context_init_ssl_library (ssl.c:148)
==30484==    by 0x416F9E: lws_create_context (context.c:1272)
==30484==    by 0x40706E: create_context (test-limit.c:323)
==30484==    by 0x4074D6: main (test-limit.c:413)
==30484== 
==30484== 32 bytes in 1 blocks are still reachable in loss record 7 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x50F5DB2: CRYPTO_malloc (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30484==    by 0x517BECC: sk_new (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30484==    by 0x4E745A9: ??? (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30484==    by 0x4E76528: SSL_COMP_get_compression_methods (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30484==    by 0x4E7B8A8: SSL_library_init (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30484==    by 0x41EAA2: lws_context_init_ssl_library (ssl.c:148)
==30484==    by 0x416F9E: lws_create_context (context.c:1272)
==30484==    by 0x40706E: create_context (test-limit.c:323)
==30484==    by 0x4074D6: main (test-limit.c:413)
==30484== 
==30484== 152 (56 direct, 96 indirect) bytes in 1 blocks are definitely lost in loss record 8 of 8
==30484==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x4C2CF1F: realloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30484==    by 0x41813E: _realloc (alloc.c:64)
==30484==    by 0x4181B1: lws_zalloc (alloc.c:82)
==30484==    by 0x41C1DB: lws_client_connect_via_info (client-handshake.c:801)
==30484==    by 0x40731D: create_request (test-limit.c:376)
==30484==    by 0x4075C4: main (test-limit.c:444)
==30484== 
==30484== LEAK SUMMARY:
==30484==    definitely lost: 56 bytes in 1 blocks
==30484==    indirectly lost: 96 bytes in 5 blocks
==30484==      possibly lost: 0 bytes in 0 blocks
==30484==    still reachable: 64 bytes in 2 blocks
==30484==         suppressed: 0 bytes in 0 blocks
==30484== 
==30484== For counts of detected and suppressed errors, rerun with: -v
==30484== ERROR SUMMARY: 1 errors from 1 contexts (suppressed: 0 from 0)
```

**Comment by freemine** [comment]
And after patching:
```
==30734== 
==30734== HEAP SUMMARY:
==30734==     in use at exit: 64 bytes in 2 blocks
==30734==   total heap usage: 3,394 allocs, 3,392 frees, 438,780 bytes allocated
==30734== 
==30734== 32 bytes in 1 blocks are still reachable in loss record 1 of 2
==30734==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30734==    by 0x50F5DB2: CRYPTO_malloc (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30734==    by 0x517BEAE: sk_new (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30734==    by 0x4E745A9: ??? (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30734==    by 0x4E76528: SSL_COMP_get_compression_methods (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30734==    by 0x4E7B8A8: SSL_library_init (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30734==    by 0x41EAA2: lws_context_init_ssl_library (ssl.c:148)
==30734==    by 0x416F9E: lws_create_context (context.c:1272)
==30734==    by 0x40706E: create_context (test-limit.c:323)
==30734==    by 0x4074D6: main (test-limit.c:413)
==30734== 
==30734== 32 bytes in 1 blocks are still reachable in loss record 2 of 2
==30734==    at 0x4C2AB80: malloc (in /usr/lib/valgrind/vgpreload_memcheck-amd64-linux.so)
==30734==    by 0x50F5DB2: CRYPTO_malloc (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30734==    by 0x517BECC: sk_new (in /lib/x86_64-linux-gnu/libcrypto.so.1.0.0)
==30734==    by 0x4E745A9: ??? (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30734==    by 0x4E76528: SSL_COMP_get_compression_methods (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30734==    by 0x4E7B8A8: SSL_library_init (in /lib/x86_64-linux-gnu/libssl.so.1.0.0)
==30734==    by 0x41EAA2: lws_context_init_ssl_library (ssl.c:148)
==30734==    by 0x416F9E: lws_create_context (context.c:1272)
==30734==    by 0x40706E: create_context (test-limit.c:323)
==30734==    by 0x4074D6: main (test-limit.c:413)
==30734== 
==30734== LEAK SUMMARY:
==30734==    definitely lost: 0 bytes in 0 blocks
==30734==    indirectly lost: 0 bytes in 0 blocks
==30734==      possibly lost: 0 bytes in 0 blocks
==30734==    still reachable: 64 bytes in 2 blocks
==30734==         suppressed: 0 bytes in 0 blocks
==30734== 
==30734== For counts of detected and suppressed errors, rerun with: -v
==30734== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 0 from 0)
```

**Comment by lws-team** [comment]
This should be solved by the patch on master e02a999863750cf71335fd9a2cb303188178716c.  Please give it a try.

The openssl "still reachable" are not actually leaks but just that a pointer somewhere in openssl still points to an allocation it freed.

**Comment by freemine** [comment]
Well, with this patch, if you does not call lws_set_timeout to kill that wsi instance synchronously before calling lws_destroy_context, there's no leakage.

But, if you call lws_set_timeout, that wsi instance will be `freed` twice. You can get the report from
valgrind. I tried to trace the code and found out this clue (please correct me if I am wrong):
1. When max_http_header_pool got hit, the new connection request will be put on a so-called waiting-list;
2. Next time when lws_service is running, it'll check that waiting-list, and find out the wsi, trying to
    apply connect_2.
3. If at this moment, getaddrinfo failed, we go into oom4. We clear the resouces related to wsi.
4. But we failed to remove wsi from the waiting-list. This leave a clue to some api to trace that
   dangling wsi out.

I'll show some code snippet here:
lws_client_connect_via_info:
```
838   /* if we went on the waiting list, no probs just return the wsi
 839    * when we get the ah, now or later, he will call
 840    * lws_client_connect_via_info2() below.
 841    */
 842   if (lws_header_table_attach(wsi, 0) < 0) {
 843     /*
 844      * if we failed here, the connection is already closed
 845      * and freed.
 846      */
 847     goto bail2;
 848   }
```
lws_header_table_attach:
```
 249   if (n) {
 250     /*
 251      * Pool is either all busy, or we don't want to give this
 252      * particular guy an ah right now...
 253      *
 254      * Make sure we are on the waiting list, and return that we
 255      * weren't able to provide the ah
 256      */
 257     _lws_header_ensure_we_are_on_waiting_list(wsi);
 258 
 259     goto bail;
 260   }
```
lws_close_free_wsi:
```
 857 
 858   lws_close_free_wsi_final(wsi);
 859 }
```
lws_close_free_wsi_final:
```
 898 
 899   lws_free_wsi(wsi);
 900 }
```
lws_free_wsi:
```
  84   /* we may not have an ah, but may be on the waiting list... */
  85   lwsl_info("ah det due to close\n");
  86   /* we're closing, losing some rx is OK */
  87   lws_header_table_force_to_detachable_state(wsi);
  88   lws_header_table_detach(wsi, 0);
```
lws_header_table_detach:
```
 339   lws_pt_lock(pt);
 340   __lws_remove_from_ah_waiting_list(wsi);
 341   lws_pt_unlock(pt);
```

Patching:
Although we can replace lws_free(wsi) with lws_close_free_wsi_final(wsi), but we'll suffer losing Connection_closed event to notify.
So I still suggest replacing lws_free(wsi) with lws_close_free_wsi(wsi), except that this function is
a huge stuff I am not very much clear about.

**Comment by lws-team** [comment]
Thanks.  So the error path leak is solved.

> But, if you call lws_set_timeout, that wsi instance will be freed twice.

Is it freed twice if we don't run your routine before the context destroy?

**Comment by freemine** [comment]
```
Is it freed twice if we don't run your routine before the context destroy?
```
If you mean lws_close_free_wsi(wsi, LWS_CLOSE_STATUS_NOSTATUS, "client_connect2");
No, freed only once, no abortion

If you mean lws_client_stash_destroy(wsi);lws_free(wsi);, 
Yes, freed twice, and abortion.  I guess waiting-list might not be traversed during lws_context_destroy
call, I am not very sure.

**Comment by freemine** [comment]
So https://github.com/warmcat/libwebsockets/commit/e02a999863750cf71335fd9a2cb303188178716c suffers 'double-free' pain.

**Comment by lws-team** [comment]
Okay... so, how do I reproduce this double-free using the test client?

**Comment by freemine** [comment]
[test-limit.c.txt](https://github.com/warmcat/libwebsockets/files/1758391/test-limit.c.txt)

just valgrind --leak-check=full --show-leak-kinds=all <path-to>/libwebsockets-test-limit

There are some notes in the file on the very top of the code. Hope this will help.

**Comment by lws-team** [comment]
I mean the lws test client, without your changes.

**Comment by freemine** [comment]
No, I guess there's no way to reproduce with test-client without any change.
Because that 'double-free' only exists for multiple-requests. I don't know how I can make different requests at the same time in test-client. Can I?

**Comment by lws-team** [comment]
I don't know why multiple requests would lead to any double-free, but sure, you can just run the test server in one terminal and then `libwebsockets-test-client -m` to open three client simultaneous client connections back to the server.  Valgrind says no problem.

**Comment by freemine** [comment]
Ok, that 'double-free' only happens on client side with concurrent requests to what ever server is.
So, can not expect to find out in server side.

**Comment by freemine** [comment]
test-limit.c that I use to generate 'double-free' is just a client-side app using lws to make several http requests to outside server (no matter it's lws back-ended or nginx, or what ever it is), and 'double-free' occurs in that test-limit process.

**Comment by freemine** [comment]
Please forget about that 'unacceptable' phrase. :)
I happend to use lws in my server program, but only use lws's client-side apis, not server-side api.

**Comment by lws-team** [comment]
The difficulty is the last time I looked at your code, the problem was in there, not lws.  I appreciate your helping out with the error path leak, which was a separate issue.  But I only want to support lws, and fix problems there, not take on responsibility for third-party code.

It's not impossible there is another problem in lws here.  But it would be great if it could be triggered on the test client, perhaps with a small hack patch to make it easy to reproduce and clear what is needed to provoke it.

Btw don't get confused about running the test server.  It's only necessary to give the test client something to connect to with the right ws protocol.  It's testing the test client.

**Comment by freemine** [comment]
```
But I only want to support lws, and fix problems there, not take on responsibility 
for third-party code.
```
Agree.

I have no intention to deny any criticize upon that demo app (test-limit). test-limit.c is there only because I failed to find test-client could help to reproduce that problem. test-limit.c is not my server program, it's just a small demo.

I am thinking if i could help out with `a small hack patch` to test-client as you just suggested, but I am kinda nervous, because in fact after all, that I would once again add some-hand-made-codes in test-client almost same as those in test-limit. That just won't help, don't you think so?

Please let me know, is there any other way to proceed.

**Comment by lws-team** [comment]
I dunno, you're saying 

> So e02a999 suffers 'double-free' pain.

I don't see how that patch can double-free anything since the function it calls sets every pointer it frees to NULL.  And the free function is a NOP if passed NULL.

I tried things like client connects, connects to fake dns name, failed connects, multiple connects (you thought this would trigger it) in the test client, no problem.  If it only happens in your app, how are we certain it's not provoked by something your app does?  The original issue was caused by your app keeping wsi pointers and closing them again after they had already been destroyed.

**Comment by freemine** [comment]
Have you tried to call lws_set_timeout just before lws_context_destroy?

If it's ok, could you please give me your new version of that test-client, maybe we can proceed in this way?

```
If it only happens in your app, how are we certain it's not provoked by something your app does? 
```
Well, I'll double-check my code.


BTW, let's say if max_http_header_pool was set 1,
steps as follows:
open 1 connection to http://www.microsoft.com
open 1 connection to http://somewhere.that.does.not.exist
lws_service
lws_set_timeout(wsi); // wsi belongs to somewhere.that.does.not.exist.

**Comment by freemine** [comment]
pseudocode:
with max_http_header_pool set to `1`
1. create request to http://www.microsoft.com
2. create request to http://somewhere.that.does.not.exist, and store wsi locally
3. call lws_service(context, 3000);
4. call lws_service(context, 3000);
5. call lws_set_timeout(wsi,.....,kill_sync);    // here, double-free'd

you can test with you code, if you can not reproduce, I'll double-check myself again.

**Comment by freemine** [comment]
```
I don't see how that patch can double-free anything since the function it calls sets every pointer it frees to NULL. And the free function is a NOP if passed NULL.
```
just a little remind: have you processed 'waiting list', and remove wsi out?

**Comment by freemine** [comment]
```
The original issue was caused by your app keeping wsi pointers and closing them again after they had already been destroyed.
```
I just give a quick glance in lws source code, with https://github.com/warmcat/libwebsockets/commit/e02a999863750cf71335fd9a2cb303188178716c,
waiting-list has been also cleaned, so there's no leak any more in this situation.

The only difference is, in that situation, callers will never be able to get WSI_DESTROYED event. Is it?
If so, how could I know in what circumstance cce is the last event, and user shall not wait until WSI_DETROY. Because as usually, failed connections will trigger cce and WSI_DESTROY.

**Comment by lws-team** [comment]
> have you processed 'waiting list', and remove wsi out?

oom4 already handles removing the wsi from the ah waiting list

    lws_header_table_detach(wsi, 0);

this calls through to

    __lws_remove_from_ah_waiting_list();

If this new scenario happens after the fds insertion, ie, after the time oom4 is the correct error path, it will handle it by doing the full close flow, which is what you are saying "fixes" it.


> call lws_set_timeout(wsi,.....,kill_sync); // here, double-free'd

Thanks, this is basically what I was looking for by asking for the hack on the test client, to see the minimum you are adding to make the problem come.

At this time, after you called service, you cannot casually refer to your copy of the original client wsi any more.  During service, it may have failed the connect and been deleted.

If you look at the test client, he takes care to set the client wsi to NULL on CLOSE or CONNECTION_ERROR, so your code can understand that happened and cannot defreference the dead pointer.  You also need to do this.

**Comment by freemine** [comment]
Thanks. Seems like we are about to get to the point.

Let me spend a little time to address my points here, and please correct me if I am wrong anyway, besides, give us a solid guidance, Thx.


Normally, users can use different two ways to bind their heap resource to wsi:
```
1. set lws_client_connect_info.userdata when calling lws_client_connect_via_info,
    expecting that'll be get via callback function's user parameter;
2. manage a kv set, to map wsi as key to user-space-data as value. 
```
```
Simply, method 1 overwhelms method 2 in performance,
because of no extra search on the set for key wsi. 
```
Thus, I'd like to discuss method 1 in detail here after.

Quote: 
```
"You're misunderstanding something, CLIENT_CONNECTION_ERROR implies lws is 
in the process of closing and freeing that wsi. 
As it is, what you think is a list of live wsi because you add them to it,
is oudated and contains pointers to freed wsi because you don't 
track CLIENT_CONNECTION_ERROR. That's not a problem in lws"
```

I am thinking as suggested, and would like to raise questions in case when cce happened, 
```
option 1:
user-side code free heap resource, because lws is in the process of closing 
and freeing that wsi?
But, how about those failed connections that still continue to generate 
LWS_CALLBACK_CLOSED_HTTP/LWS_CALLBACK_WSI_DESTROY sort of event(s), 
with user parameter set to non-NULL in callback function, how do we know it has 
been cce'd or not? Will I risk in reading 'dangling' pointer or ultimately 'double-free' 
problem?
(fact: in that rare case stated early in this thread of issue discussion, after cce of 
that wsi, no events get called. Thus, totally different event-ending-chain)

option 2:
user-side code does nothing, because lws will clear wsi in it's own way?
But when shall use-space-data on heap get the chance to free? Only after
lws_context_detroy call? Which usually denotes the ending of process.
What if it's a long-running program and continues to service these kind of 
client connection requests, wait until Out-Of-Memory?
```

**Please give us a guidance or a clue, thank you very much**

BTW, to public who are watching this discussion, test-limit.c is just a what-so-ever-called 
bug-reproducer, although there are codes following lws_set_timeout call, they won't 
get called when error occurs, all I count on is logs prior to that line, or report from valgrind.
Just a little trick to keep lines in test-limit.c as little as possible :) 
As what `lws-team` just mentioned, that is NOT recommended practice in real world project, 
please do NOT follow.

**Comment by lws-team** [comment]
Well, it means we can discard your various speculations above about what may be wrong in lws (...except the error path, which was real and got fixed already).  The problem you had is related to keeping an invalidated pointer.

There are two main points to understand... the first I explained before, unlike server connections which are instantiated as the result of an accept() and so are born already established on both sides, client connections are less certain.  First they have to get through the name resolution, and only then if successful try to touch the peer; the connect() completes or fails asynchronously.  So before there is any kind of client connection establishment, at tcp or any other level, the attempt has many additional ways to fail.  These extra ways to fail before anything got done are handled by CONNECTION_ERROR.

For either kind of connection though, the **peer may choose to close it at any time**, for that reason you are in the business of waiting to be informed about the disposition of connections, you are not in control of what the other side decided to do, dog chewed the ethernet cable or whatever.  You also may choose to close it at any time.  That will be notified by one or other flavour of _CLOSE.

About binding external data to the wsi

```
struct lws_client_connect_info {
	...
	void *userdata;
        /**< if non-NULL, use this as wsi user_data instead of malloc it */
```

**Comment by freemine** [comment]
Thank you very much for such a detailed explanation.

Just curiously one last question, in any of such CONNECTION_ERROR event, when wsi is free'd internally by lws, for what purpose, does lws lib choose NOT to trigger WSI_DESTROY to user-side
code via callback function？I am just wondering if WSI_DESTROY is triggered for what ever reason it is when wsi is going to be free'd, it'll be easy for user-side code to cope with their 'userdata'.

**Comment by lws-team** [comment]
Because the corresponding wsi creation callback WSI_CREATE is not issued until after the wsi connect() completed successfully, and we got past the code that uses the 'oom4' 'lite close' for cleanup.  Lws shouldn't issue WSI_DESTROY on a wsi that never issued WSI_CREATE on, it would be unbalanced.

You can indeed use WSI_CREATE and WSI_DESTROY to track lifecycle, but they will only occur on client connections that were successful enough to get connected.  (server wsi as explained before only get created as connected wsi so for those every server wsi gets a WSI_CREATE and WSI_DESTROY).  To know the outcome for all client wsi you have to monitor CONNECTION_ERROR and CLOSE.

**Comment by freemine** [comment]
Ok, that makes sense, cauze WSI_CREATE and WSI_DESTROY shall be paired.
I had took it for granted that when lws_client_connect_via_info returns a valid wsi means WSI_CREATE has been triggered. My fault.

This issue closed then. Thank you very much!
